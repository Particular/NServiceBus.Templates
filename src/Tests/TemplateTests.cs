using System;
using System.IO;
using System.Linq;
using ApprovalTests;
using NUnit.Framework;

[TestFixture]
public class TemplateTests : IDisposable
{
    public TemplateTests()
    {
        Uninstall();
        Install();
    }

    static void Install()
    {
        //dotnet new --install \nugets\NServiceBus.Template.WindowsService.xxx.nupkg
        var nugets = Path.GetFullPath(Path.Combine(ProjectDirectory.ProjectPath, "../../nugets"));
        var nugetPath = Directory.EnumerateFiles(nugets, "*.nupkg").Single();
        DotNetTemplatesHelper.Install(nugetPath);
    }

    void Uninstall()
    {
        // dotnet new --uninstall ParticularTemplates
        DotNetTemplatesHelper.Uninstall("ParticularTemplates");
    }

    public void Dispose()
    {
        Uninstall();
    }

    [Test]
    public void WindowsService()
    {
        //dotnet new nsbservice
        var targetDirectory = ProjectDirectory.GetSandboxPath(nameof(WindowsService));
        VerifyAndBuild(targetDirectory);
    }

    [Test]
    public void ScAdapterService()
    {
        //dotnet new scadapterservice
        var targetDirectory = ProjectDirectory.GetSandboxPath(nameof(ScAdapterService));
        VerifyAndBuild(targetDirectory);
    }

    static void VerifyAndBuild(string targetDirectory)
    {
        DotNetTemplatesHelper.Run("nsbservice", targetDirectory);
        VerifyDirectory(targetDirectory);
        DotNetTemplatesHelper.Build(targetDirectory);
    }

    static void VerifyDirectory(string targetDirectory)
    {
        var files = Directory.EnumerateFiles(targetDirectory, "*.*");

        var dictionary = files.ToDictionary(
            keySelector: x => Path.GetFileName(x),
            elementSelector: x => $"\r\n{File.ReadAllText(x)}");
        Approvals.VerifyAll(dictionary);
    }
}