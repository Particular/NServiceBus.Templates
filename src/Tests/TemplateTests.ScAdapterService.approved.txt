ProgramService.cs => 
using System;
using System.ComponentModel;
using System.Linq;
using System.ServiceProcess;
using System.Threading.Tasks;
using NServiceBus;
using NServiceBus.Logging;
using ServiceControl.TransportAdapter;

[DesignerCategory("Code")]
class ProgramService : ServiceBase
{
    const string runAsServiceArg = "--run-as-service";

    ITransportAdapter adapter;

    static ILog logger;

    static ProgramService()
    {
        //TODO: optionally choose a custom logging library
        // https://docs.particular.net/nservicebus/logging/#custom-logging
        // LogManager.Use<TheLoggingFactory>();
        logger = LogManager.GetLogger<ProgramService>();
    }

    static void Main(string[] args)
    {
        using (var service = new ProgramService())
        {
            // pass argument at command line to run as a windows service
            if (args.Contains(runAsServiceArg))
            {
                Run(service);
                return;
            }

            Console.Title = "ScAdapterService";
            Console.CancelKeyPress += (sender, e) => { service.OnStop(); };
            service.OnStart(null);
            Console.WriteLine("\r\nPress enter key to stop program\r\n");
            Console.Read();
            service.OnStop();
        }
    }

    protected override void OnStart(string[] args)
    {
        AsyncOnStart().GetAwaiter().GetResult();
    }

    async Task AsyncOnStart()
    {
        try
        {
            var adapterConfig = new TransportAdapterConfig<LearningTransport, LearningTransport>("TransportAdapter.WindowsService");

            adapterConfig.CustomizeEndpointTransport(t =>
            {
                //TODO: Customize the endpoint-facing side of the adapter
                //Use exactly same settings as in regular endpoints
            });

            adapterConfig.CustomizeServiceControlTransport(t =>
            {
                //TODO: Customize the ServiceControl-facing side of the adapter
                //e.g. specify the same connection string as ServiceControl uses.
            });

            adapter = TransportAdapter.Create(adapterConfig);

            await adapter.Start().ConfigureAwait(false);
        }
        catch (Exception exception)
        {
            Exit("Failed to start", exception);
        }
    }

    void Exit(string failedToStart, Exception exception)
    {
        logger.Fatal(failedToStart, exception);

        //TODO: When using an external logging framework it is important to flush any pending entries prior to calling FailFast
        // https://docs.particular.net/nservicebus/hosting/critical-errors#when-to-override-the-default-critical-error-action
        Environment.FailFast(failedToStart, exception);
    }

    protected override void OnStop()
    {
        adapter?.Stop().GetAwaiter().GetResult();
    }
}


ScAdapterService.csproj => 
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net462</TargetFramework>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>

  <ItemGroup>
    <Reference Include="System.ServiceProcess" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="NServiceBus" Version="7.0.0-*" />
    <PackageReference Include="NServiceBus.Raw" Version="3.0.0-*" />
    <PackageReference Include="ServiceControl.TransportAdapter" Version="2.0.0-*" />
  </ItemGroup>

</Project>


